lib/fluent/plugin/out_forward/socket_cache.rb

```diff
@@ -19,135 +19,118 @@
 module Fluent::Plugin
   class ForwardOutput < Output
     class SocketCache
-      TimedSocket = Struct.new(:timeout, :sock, :ref)
+      TimedSocket = Struct.new(:timeout, :key, :sock)
 
       def initialize(timeout, log)
         @log = log
         @timeout = timeout
-        @active_socks = {}
-        @inactive_socks = {}
+        @available_sockets = Hash.new { |obj, k| obj[k] = [] }
+        @inflight_sockets = {}
+        @inactive_sockets = []
         @mutex = Mutex.new
       end
 
-      def revoke(key = Thread.current.object_id)
+      def checkout_or(key)
         @mutex.synchronize do
-          if @active_socks[key]
-            @inactive_socks[key] = @active_socks.delete(key)
-            @inactive_socks[key].ref = 0
+          tsock = pick_socket(key)
+
+          unless tsock
+            val = yield
+            new_tsock = TimedSocket.new(timeout, key, val)
+            @log.debug("connect new socket #{new_tsock}")
+
+            @inflight_sockets[val] = new_tsock
+            return new_tsock.sock
           end
+          tsock.sock
         end
       end
 
-      def clear
+      def checkin(sock)
         @mutex.synchronize do
-          @inactive_socks.values.each do |s|
-            s.sock.close rescue nil
+          if (s = @inflight_sockets.delete(sock))
+            @available_sockets[s.key] << s
+          else
+            @log.debug("there is no socket #{sock}")
           end
-          @inactive_socks.clear
+        end
+      end
 
-          @active_socks.values.each do |s|
-            s.sock.close rescue nil
+      def revoke(sock)
+        @mutex.synchronize do
+          if (s = @inflight_sockets.delete(sock))
+            @inactive_sockets << s
+          else
+            @log.debug("there is no socket #{sock}")
           end
-          @active_socks.clear
         end
       end
 
       def purge_obsolete_socks
+        sockets = []
+
         @mutex.synchronize do
-          @inactive_socks.keys.each do |k|
-            # 0 means sockets stored in this class received all acks
-            if @inactive_socks[k].ref <= 0
-              s = @inactive_socks.delete(k)
-              s.sock.close  rescue nil
-              @log.debug("purged obsolete socket #{s.sock}")
+          # don't touch @inflight_sockets
+
+          @available_sockets.each do |_, socks|
+            socks.each do |sock|
+              if expired_socket?(sock)
+                sockets << sock
+                socks.delete(sock)
+              end
             end
           end
+          @available_sockets = @available_sockets.select { |_, v| !v.empty? }
 
-          @active_socks.keys.each do |k|
-            if expired?(k) && @active_socks[k].ref <= 0
-              @inactive_socks[k] = @active_socks.delete(k)
-            end
-          end
+          sockets += @inactive_sockets
+          @inactive_sockets.clear
+        end
+
+        while (s = sockets.pop)
```

Comment:
`each` is more lightweight due to no `sockets` update.

Reply:
right……… fixed https://github.com/fluent/fluentd/pull/2501/commits/eda267cee62e3230d700037fef0fd90a3076c5fa